---
title: "SystemGenomicsAnalysis"
author: "Moritz Ullhofen"
date: "`r Sys.Date()`"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(BiocManager)
BiocManager::install("biomaRt")
library(biomaRt)
library(ggplot2)
library(cowplot)
library(here)
BiocManager::install("DESeq2")
BiocManager::install("sva")
BiocManager::install("genefilter")
BiocManager::install("biomaRt")
BiocManager::install("ggrepel")
BiocManager::install("DESeq2")
BiocManager::install("Biostrings")
BiocManager::install("tximport")
library(tximport)
library(biomaRt)
library(sva)
library(pbapply)
library(devtools)
library(Biostrings)
devtools::install_github("slowkow/ggrepel")
library(ggrepel)
library(DESeq2)

```

# 3_2 Import data to R

```{r}
samples <- list.files(here("EulerDownload","rsem"))

expr <- sapply(samples, function(sample){
  file <- paste0(here("EulerDownload","rsem"),"/",sample,"/",sample,".genes.results")
  quant <- read.csv(file, sep="\t", header=T)
  tpm <- setNames(quant$TPM, quant$gene_id)
  return(tpm)
})

```

#Read in metadata
```{r}
meta <- read.csv(here("additional_meta.csv"), header=T) %>%
  inner_join(read.csv(here("SraRunTable.txt"), header=T),
             by ="GEO_Accession..exp.",
             suffix = c("",".y"))
meta$X <- NULL
expr <- expr[,meta$Run]
```

```{r}

expr <- expr[,meta$Run]

#more gene information
ensembl <- useEnsembl(biomart = "ensembl",
                      dataset = "hsapiens_gene_ensembl")
meta_genes <- getBM(attributes = c("ensembl_gene_id",
                                   "ensembl_gene_id_version",
                                   "hgnc_symbol",
                                   "description",
                                   "chromosome_name",
                                   "start_position",
                                   "end_position",
                                   "strand"),
                    filters = "ensembl_gene_id_version",
                    values = rownames(expr),
                    mart = ensembl) %>%
  right_join(data.frame(ensembl_gene_id_version = rownames(expr)),
             by = "ensembl_gene_id_version") %>%
  distinct(ensembl_gene_id_version, .keep_all = TRUE)
expr <- expr[meta_genes$ensembl_gene_id_version,]

```

# 3-3 Comparison of transcriptomic profiles across samples

```{r}
#check dimensions, and average expression level
dim(expr)

avg_expr <- rowMeans(expr)
layout(matrix(1:2, nrow=1))
hist(avg_expr)
hist(log10(avg_expr + 1))

#log tranform y-axis
ggplot(data.frame(avg_expr), aes(x=avg_expr)) +
  geom_histogram(bins = 50) +
  scale_x_continuous(breaks = c(0,1,10,100,1000,10000,20000), trans="log1p", expand=c(0,0)) +
  scale_y_continuous(breaks = c(0,1), expand=c(0,0), trans="log1p") +
  theme_minimal()

#how many samples each gene detected
num_det <- rowSums(expr > 0)
hist(num_det)
```
```{r}
#mark unexpressed or extremely low expressed genes
expressed <- rowMeans(expr > 0) >= 0.5 | rowMeans(expr) >= 1
meta_genes$expressed <- expressed

#see the correlation between the genes
corr_pearson <- cor(log1p(expr[meta_genes$expressed,]))
corr_spearman <- cor(expr[meta_genes$expressed,], method = "spearman")

#hierarchical clusterring
hcl_pearson <- hclust(as.dist(1 - corr_pearson))
hcl_spearman <- hclust(as.dist(1 - corr_spearman))

layout(matrix(1:2,nrow=1))
plot(hcl_pearson)
plot(hcl_spearman)

#Example of plotting with different labels in spearman case
layout(matrix(1:2,nrow=1))
plot(hcl_spearman, labels = meta$Individual)
plot(hcl_spearman, labels = meta$Sex) #maybe choose other variable (disease_state?)
plot(hcl_spearman, labels = meta$disease_state)
```


```{r}
#Dimension reduced with PCA
pca <- prcomp(log1p(t(expr[meta_genes$expressed,])), center = T, scale. = T)
eigs <- pca$sdev^2
plot(1:length(eigs), eigs/sum(eigs)) # i think this is normalized now but i am not sure
```

```{r}
#Visualize first two pcas
ggplot(data.frame(pca$x, meta)) +
geom_point(aes(x = PC1, y = PC2, color = Sex, shape = disease_state), size = 5)
```

```{r}
#Identify highly variable genes
estimate_variability <- function(expr){
means <- apply(expr, 1, mean)
vars <- apply(expr, 1, var)
cv2 <- vars / means^2
minMeanForFit <- unname(median(means[which(cv2 > 0.3)]))
useForFit <- means >= minMeanForFit
fit <- glm.fit(x = cbind(a0 = 1, a1tilde = 1/means[useForFit]),
y = cv2[useForFit],
family = Gamma(link = "identity"))
a0 <- unname(fit$coefficients["a0"])
a1 <- unname(fit$coefficients["a1tilde"])
xg <- exp(seq(min(log(means[means>0])), max(log(means)), length.out=1000))
vfit <- a1/xg + a0
df <- ncol(expr) - 1
afit <- a1/means+a0
varFitRatio <- vars/(afit*means^2)
pval <- pchisq(varFitRatio*df,df=df,lower.tail=F)
res <- data.frame(mean = means,
var = vars,
cv2 = cv2,
useForFit = useForFit,
pval = pval,
padj = p.adjust(pval, method="BH"),
row.names = rownames(expr))
return(res)
}

var_genes <- estimate_variability(expr[meta_genes$expressed,])
meta_genes$highvar <- meta_genes$ensembl_gene_id_version %in% rownames(var_genes)[which(var_genes$padj < 0.01)] 

#Repeat the clustering and pca with highly variable genes
corr_spearman_highvar <- cor(expr[meta_genes$highvar,], method = "spearman")
hcl_spearman_highvar <- hclust(as.dist(1 - corr_spearman_highvar))
layout(matrix(1:2,nrow=1))

plot(hcl_spearman_highvar, labels = meta$Sex)
plot(hcl_spearman_highvar, labels = meta$disease_state)

pca_highvar <- prcomp(log1p(t(expr[meta_genes$highvar,])), center = TRUE, scale. = TRUE)
ggplot(data.frame(pca_highvar$x, meta)) +
geom_point(aes(x = PC1, y = PC2, color = Sex, shape = disease_state), size = 5)
```

```{r}
#Batch effect correction
expr_combat<-ComBat_seq(counts=expr,batch=meta$disease_state)

corr_spearman_combat <- cor(expr_combat[meta_genes$expressed,], method = "spearman")
hcl_spearman_combat <- hclust(as.dist(1 - corr_spearman_combat))
layout(matrix(1:2,nrow=1))

plot(hcl_spearman_combat, labels = meta$Sex)
plot(hcl_spearman_combat, labels = meta$disease_state)

pca_combat <- prcomp(log1p(t(expr_combat[meta_genes$expressed,])), center = TRUE, scale. = TRUE)
ggplot(data.frame(pca_combat$x, meta)) +
geom_point(aes(x = PC1, y = PC2, color = Sex, shape = disease_state), size = 5)

```

```{r}
#ANOVA test between the different sexes with bonferoni correction for multiple testing
De_test<- function(expr,
                    cond,
                    ctrl = NULL,
                    covar = NULL,
                    padj_method = "Bonferroni"){
  pval_fc <- data.frame(t(pbapply(expr, 1, function(e){
    dat <- data.frame(y = log1p(e),
                      cond = cond)
    if (! is.null(covar))
      dat <- data.frame(dat, covar)
    m1 <- lm(y ~ ., data = dat)
    m0 <- lm(y ~ . - cond, data = dat)
    test <- anova(m1, m0)
    pval <- test$Pr[2]
    avgs <- tapply(log1p(e), cond, mean)
    if (! is.null(ctrl) && sum(cond %in% ctrl) > 0){
      fc <- exp(max(avgs[names(avgs) != ctrl]) - avgs[ctrl])
    } else{
      fc <- exp(max(avgs) - min(avgs))
    }
    return(c(pval = unname(pval), fc = unname(fc)))
  })), row.names = rownames(expr))
  padj <- p.adjust(pval_fc$pval, method = padj_method)
  return(data.frame(pval_fc, "padj" = padj)[,c("pval","padj","fc")])
}
res_DE <- DE_test(expr = expr[meta_genes$expressed,],
                  cond = meta$disease_state, 
                  ctrl =  filter(meta, disease_state=="Control")) %>%
  tibble::rownames_to_column("gene")

```

```{r}
#vulcano plot for logfold change -> check if gene labels are mapped correctly
res_DE <- res_DE %>%
  mutate(DE = padj < 0.1 & fc > 2) %>%
  mutate(DEG = ifelse(DE, meta_genes$hgnc_symbol, NA))
ggplot(res_DE, aes(x = log(fc), y = -log10(padj), col=DE, label=DEG)) +
  geom_point() +
  geom_text_repel() +
  geom_vline(xintercept=c(log(2), 0), col="#303030", linetype="dotted") +
  geom_hline(yintercept=-log10(0.1), col="#303030", linetype="dotted") +
  scale_color_manual(values=c("#909090", "red")) +
  theme_minimal()

```
```{r}
#DE analysis with DESeq2

ensembl <- useEnsembl(biomart = "ensembl",
                      dataset = "hsapiens_gene_ensembl")
tx2gene <- getBM(attributes = c("ensembl_transcript_id_version",
                                "ensembl_gene_id_version"),
                 filters = "ensembl_gene_id_version",
                 values = rownames(expr),
                 mart = ensembl) %>%
  dplyr::select(ensembl_transcript_id_version, ensembl_gene_id_version)

samples <-  list.files(here("EulerDownload", "rsem"))
files <- file.path(here("EulerDownload", "rsem"), samples, paste0(samples,".isoforms.results"))
txi <- tximport(files, type = "rsem", tx2gene = tx2gene)

dds <- DESeqDataSetFromTximport(txi,
                                colData = meta,
                                design = ~ disease_state)

dds_filtered <- dds[intersect(rownames(expr)[meta_genes$expressed], rownames(dds)),]
dds_filtered <- DESeq(dds_filtered, test="Wald")
res_DESeq2 <- results(dds_filtered)

cor(res_DESeq2$padj,
    res_DE %>% filter(gene %in% rownames(res_DESeq2)) %>% pull(padj),
    method="spearman", use="complete.obs")

layout(matrix(1:2, nrow=1))
plot(-log10(res_DESeq2$pvalue),
     -log10(res_DE %>% filter(gene %in% rownames(res_DESeq2)) %>% pull(pval)),
     xlab = "-log10(pval DESeq2)", ylab = "-log10(pval DE)", pch=16)
smoothScatter(-log10(res_DESeq2$pvalue),
              -log10(res_DE %>% filter(gene %in% rownames(res_DESeq2)) %>% pull(pval)),
              xlab = "-log10(pval DESeq2)", ylab = "-log10(pval DE)", pch=16)

```


