---
title: "SystemGenomicsAnalysis"
author: "Moritz Ullhofen"
date: "`r Sys.Date()`"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(BiocManager)
library(biomaRt)
library(ggplot2)
library(cowplot)
library(here)
```

# 3_2 Import data to R

```{r}
samples <- list.files(here("EulerDownload","rsem"))

expr <- sapply(samples, function(sample){
  file <- paste0(here("EulerDownload","rsem"),"/",sample,"/",sample,".genes.results")
  quant <- read.csv(file, sep="\t", header=T)
  tpm <- setNames(quant$TPM, quant$gene_id)
  return(tpm)
})

```

#Read in metadata
```{r}
meta <- read.csv(here("additional_meta.csv"), header=T) %>%
  inner_join(read.csv(here("SraRunTable.txt"), header=T),
             by ="GEO_Accession..exp.",
             suffix = c("",".y"))
meta$X <- NULL
expr <- expr[,meta$Run]
```

```{r}

expr <- expr[,meta$Run]

#more gene information
ensembl <- useEnsembl(biomart = "ensembl",
                      dataset = "hsapiens_gene_ensembl")
meta_genes <- getBM(attributes = c("ensembl_gene_id",
                                   "ensembl_gene_id_version",
                                   "hgnc_symbol",
                                   "description",
                                   "chromosome_name",
                                   "start_position",
                                   "end_position",
                                   "strand"),
                    filters = "ensembl_gene_id_version",
                    values = rownames(expr),
                    mart = ensembl) %>%
  right_join(data.frame(ensembl_gene_id_version = rownames(expr)),
             by = "ensembl_gene_id_version") %>%
  distinct(ensembl_gene_id_version, .keep_all = TRUE)
expr <- expr[meta_genes$ensembl_gene_id_version,]

```

# 3-3 Comparison of transcriptomic profiles across samples

```{r}
#check dimensions, and average expression level
dim(expr)

avg_expr <- rowMeans(expr)
layout(matrix(1:2, nrow=1))
hist(avg_expr)
hist(log10(avg_expr + 1))

#log tranform y-axis
ggplot(data.frame(avg_expr), aes(x=avg_expr)) +
  geom_histogram(bins = 50) +
  scale_x_continuous(breaks = c(0,1,10,100,1000,10000,20000), trans="log1p", expand=c(0,0)) +
  scale_y_continuous(breaks = c(0,1), expand=c(0,0), trans="log1p") +
  theme_minimal()

#how many samples each gene detected
num_det <- rowSums(expr > 0)
hist(num_det)
```
```{r}
#mark unexpressed or extremely low expressed genes
expressed <- rowMeans(expr > 0) >= 0.5 | rowMeans(expr) >= 1
meta_genes$expressed <- expressed

#see the correlation between the genes
corr_pearson <- cor(log1p(expr[meta_genes$expressed,]))
corr_spearman <- cor(expr[meta_genes$expressed,], method = "spearman")

#hierarchical clusterring
hcl_pearson <- hclust(as.dist(1 - corr_pearson))
hcl_spearman <- hclust(as.dist(1 - corr_spearman))

layout(matrix(1:2,nrow=1))
plot(hcl_pearson)
plot(hcl_spearman)

#Example of plotting with different labels in spearman case
layout(matrix(1:2,nrow=1))
plot(hcl_spearman, labels = meta$Individual)
plot(hcl_spearman, labels = meta$Sex) #ma<be choose other variable (disease_state?)
```


```{r}
#Dimension reduced with PCA
#normalize eigenvalues of x 
pca <- prcomp(log1p(t(expr[meta_genes$expressed,])), center = TRUE, scale. = TRUE)
eigs <- pca$sdev^2
plot(1:length(eigs), eigs)

```


